### PR TITLE
Fix Uniform discrete distribution

### PR DESCRIPTION
**Pull request by jantonguirao**
Signed-off-by: Joaquin Anton <janton@nvidia.com>

#### Why we need this PR?
*Pick one, remove the rest*
- It fixes a bug in Uniform distribution producing unexpected values

#### What happened in this PR?
*Fill relevant points, put NA otherwise. Replace anything inside []*
 - What solution was applied:
     *An extra ArgValue Acquire call every iteration was overriding the values pointer inside ArgValue implementation while the per_sample_values data was kept unmodified due to the assumption that the pointer won't change.*
     *As a solution, we Acquire the values only once when a non-tensor argument is given*
     *Using TryGetArgument/TryGetRepeatedArgument to avoid reallocation*
 - Affected modules and functionalities:
     *Uniform distribution and other operators using ArgValue*
 - Key points relevant for the review:
     *All*
 - Validation and testing:
     *Added niter to uniform tests*
 - Documentation (including examples):
     *N/A*

**JIRA TASK**: *[DALI-1824*

### CODE REVIEW

**Review by JanuszL — APPROVED** [review]

File: dali/pipeline/operator/arg_helper.h [hunk 1]
```diff
@@ -116,7 +116,7 @@ class ArgValue {
       DALI_ENFORCE(is_uniform(view_.shape) && expected_shape == view_.shape[0],
         make_string("Expected uniform shape for argument \"", arg_name_,
                     "\" but got shape ", view_.shape));
-    } else {
+    } else if (view_.empty()) {
```
Comment (mzient) [inline dali/pipeline/operator/arg_helper.h:119]: 1. It assumes that non-tensor data is constant (and we may change that at some point)
2. It validates the hidden (and dangerous) assumption that Acquire is no-op - I think that hoisting this logic outside of Acquire violates encapsulation of ArgValue.
3. It makes clear that the way we obtain non-tensor arguments is inefficient, as it results in reallocation of the internal vector - use TryGetArgument / TryGetRepeatedArgument which will resize and populate an existing vector.

File: dali/operators/random/uniform_distribution.h [hunk 2]
```diff
@@ -115,9 +115,9 @@ class UniformDistribution : public RNGBase<Backend, UniformDistribution<Backend>
 
   void AcquireArgs(const OpSpec &spec, const workspace_t<Backend> &ws, int nsamples) {
     if (values_.IsDefined()) {
-      values_.Acquire(spec, ws, nsamples, false);
       // read only once for build time arguments
       if (!values_.IsConstant() || per_sample_values_.empty()) {
+        values_.Acquire(spec, ws, values_.IsConstant() ? max_batch_size_ : nsamples, false);
         per_sample_values_.resize(nsamples);
         per_sample_nvalues_.resize(nsamples);
```
Comment (mzient) [inline dali/operators/random/uniform_distribution.h:122]:
```suggestion
        per_sample_values_.resize(values_.size());
        per_sample_nvalues_.resize(values_.size());
```

Comment (mzient) [reply dali/operators/random/uniform_distribution.h:122]: and similarly in the loops below

**Comment by jantonguirao** [comment]
!build

File: dali/pipeline/operator/arg_helper_test.cc [hunk 3]
```diff
@@ -83,17 +83,17 @@ TEST(ArgValue, TensorInput_3D) {
 
 TEST(ArgValueTests, Constant_0D) {
   int nsamples = 5;
-  auto spec = OpSpec("opname").AddArg("argname", 0.123f);
-  ArgValue<float, 0> arg("argname", spec);
+  auto spec = OpSpec("Erase").AddArg("shape", 0.123f);
```
Comment (jantonguirao) [inline dali/pipeline/operator/arg_helper_test.cc:86]: Note: Using TryGetArgument* required that the spec points to an existing operator.

**Review by mzient — APPROVED** [review]

**Comment by jantonguirao** [comment]
!build

**Pull request merged**